### PR TITLE
Lde/split gemini

### LINKED ISSUE #285: Split Gemini prover
All of the Gemini prover functionality is handled by the function `Gemini::reduce_prove`. This part of the protocol involves computing commitments, adding them to the transcript, computing a challenge, then performing some subsequent calculations using the challenge. Eventually we want to compute commitments in the `work_queue` to facilitate parallelization in WASM. This requires that `Gemini::reduce_prove` be broken into two separate functions. The first contains the functionality prior to the computation of the commitments, the second contains the calculations done with the challenge generated based on the commitments.

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

Split the functionality in `Gemini::reduce_prove` into two functions that are executed across two prover "rounds". This will eventually facilitate computation of the Fold polynomial commitments in the `work_queue`.

Note: This PR also removes the dependence on Gemini from the Shplonk test, making it more clear, focused and maintainable. The test `GeminiShplonkKzgWithShift` still exists for the purpose of demontrating/testing the full Gemini/Shplonk/KZG chain.

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [x] There are no circuit changes, OR a cryptographer has been assigned for review.
- [x] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [x] The branch has been rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.
- [x] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [x] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

**Comment by ledwards2225** [comment]
This branch is based off of lde/transcript (which I expect to be merged with minimal changes, if any). It will have to be rebased to master after that merge, at which point the diff will be dramatically reduced.

File: cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp [hunk 1]
```diff
@@ -100,35 +76,20 @@ template <typename Params> class MultilinearReductionScheme {
...
+     * and G(X) = ∑ⱼ ρᵏ⁺ʲ gⱼ(X), i.e. the batched unshifted polynomials and the batched to-be-shifted polynomials. This
+     * function populates the next d-1 elements with Fold_i, i = 1, ..., d-1.
      *
      */
-    static ProverOutput<Params> reduce_prove(std::shared_ptr<CK> ck,
-                                             std::span<const Fr> mle_opening_point,
-                                             const Polynomial&& batched_shifted,       /* unshifted */
-                                             const Polynomial&& batched_to_be_shifted, /* to-be-shifted */
-                                             ProverTranscript<Fr>& transcript)
+    static void compute_fold_polynomials(std::span<const Fr> mle_opening_point, auto& fold_polynomials)
```
Comment (adr1anh) [inline cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: Small nitpick, but I'm not really a fan of using arguments for return values. We could pass `batched_shifted` and `batched_to_be_shifted` by move, and return the vector of `fold_polynomials`

File: cpp/src/barretenberg/honk/proof_system/prover.cpp [hunk 2]
```diff
@@ -303,23 +303,42 @@ template <typename settings> void Prover<settings>::execute_univariatization_rou
...
-                                         transcript);
+    // Reserve space for d+1 Fold polynomials. At the end of this round, the last d-1 polynomials will
+    // correspond to Fold^(i). At the end of the full Gemini prover protocol, the first two will
+    // be the partially evaluated Fold polynomials Fold_{r}^(0) and Fold_{-r}^(0).
+    fold_polynomials.reserve(key->log_circuit_size + 1);
+    fold_polynomials.emplace_back(batched_poly_unshifted);
+    fold_polynomials.emplace_back(batched_poly_to_be_shifted);
+
+    // Compute d-1 polynomials Fold^(i), i = 1, ..., d-1.
+    Gemini::compute_fold_polynomials(sumcheck_output.challenge_point, fold_polynomials);
```
Comment (adr1anh) [inline cpp/src/barretenberg/honk/proof_system/prover.cpp]: If the function returns `fold_polynomials`, you can assign it to the member variable of the prover class.

**Review by adr1anh — COMMENTED** [review]
Looks good. Just a minor nitpick about passing `fold_polynomial` by argument instead of returning.

File: cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp (hunk 1 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: No you're right, that's better. Will make the change

Comment (ledwards2225) [reply cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: Done

File: cpp/src/barretenberg/honk/proof_system/prover.cpp (hunk 2 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/proof_system/prover.cpp]: Done

File: cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp [hunk 3]
```diff
@@ -100,35 +76,29 @@ template <typename Params> class MultilinearReductionScheme {
...
+     * @return std::vector<Polynomial>
      */
-    static ProverOutput<Params> reduce_prove(std::shared_ptr<CK> ck,
-                                             std::span<const Fr> mle_opening_point,
-                                             const Polynomial&& batched_shifted,       /* unshifted */
-                                             const Polynomial&& batched_to_be_shifted, /* to-be-shifted */
-                                             ProverTranscript<Fr>& transcript)
+    static std::vector<Polynomial> compute_fold_polynomials(std::span<const Fr> mle_opening_point,
+                                                            const Polynomial&& batched_unshifted,
+                                                            const Polynomial&& batched_to_be_shifted)
```
Comment (adr1anh) [inline cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: Move argument should not use `const`

File: cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp [hunk 4]
```diff
@@ -100,35 +76,29 @@ template <typename Params> class MultilinearReductionScheme {
...
+        // They will eventually contain the full batched polynomial A₀ partially evaluated at the challenges r,-r.
+        // This function populates the other m-1 polynomials with the foldings of A₀.
         std::vector<Polynomial> fold_polynomials;
         fold_polynomials.reserve(num_variables + 1);
-        // F(X) = ∑ⱼ ρʲ   fⱼ(X)
-        Polynomial& batched_F = fold_polynomials.emplace_back(batched_shifted);
-        // G(X) = ∑ⱼ ρᵏ⁺ʲ gⱼ(X)
-        Polynomial& batched_G = fold_polynomials.emplace_back(batched_to_be_shifted);
+
+        Polynomial& batched_F = fold_polynomials.emplace_back(batched_unshifted);     // F(X) = ∑ⱼ ρʲ   fⱼ(X)
```
Comment (adr1anh) [inline cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: I think a copy may still be occurring here. Might require `std::move`

**Review by adr1anh — COMMENTED** [review]
Move semantics

File: cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp (hunk 4 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: Yes, thank you. Done

File: cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp (hunk 3 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/pcs/gemini/gemini.hpp]: Removed

**Review by adr1anh — APPROVED** [review]

**Pull request merged**